File: python3/Graphs/bipartite_graph_validation.py
from typing import List


def bipartite_graph_validation(graph: List[List[int]]) -> bool:
    colors = [0] * len(graph)
    # Determine if each graph component is bipartite.
    for i in range(len(graph)):
        if colors[i] == 0 and not dfs(i, 1, graph, colors):
            return False
    return True

def dfs(node: int, color: int, graph: List[List[int]], colors: List[int]) -> bool:
    colors[node] = color
    for neighbor in graph[node]:
        # If the current neighbor has the same color as the current 
        # node, the graph is not bipartite.
        if colors[neighbor] == color:
            return False
        # If the current neighbor is not colored, color it with the 
        # other color and continue the DFS.
        if colors[neighbor] == 0 and not dfs(neighbor, -color, graph, colors):
            return False
    return True

File: cpp/Graphs/bipartite_graph_validation.cpp
#include <vector>

bool bipartiteGraphValidation(std::vector<std::vector<int>>& graph) {
    std::vector<int> colors(graph.size(), 0);
    // Determine if each graph component is bipartite.
    for (int i = 0; i < graph.size(); i++) {
        if (colors[i] == 0 && !dfs(i, 1, graph, colors)) {
            return false;
        }
    }
    return true;
}

bool dfs(int node, int color, std::vector<std::vector<int>>& graph, std::vector<int>& colors) {
    colors[node] = color;
    for (int neighbor : graph[node]) {
        // If the current neighbor has the same color as the current 
        // node, the graph is not bipartite.
        if (colors[neighbor] == color) {
            return false;
        }
        // If the current neighbor is not colored, color it with the 
        // other color and continue the DFS.
        if (colors[neighbor] == 0 && !dfs(neighbor, -color, graph, colors)) {
            return false;
        }
    }
    return true;
}